Bluetooth: HCI: Add option CONFIG_BUF_ACL_DATA_LEN

jostn opened the pull request:
Add the option CONFIG_BT_BUF_ACL_DATA_LEN which allows the user
to configure the max HCI ACL payload. This is needed for platorms
where the BLE LL, HCI uart and host run on three different cores.

Fixes: #30441

Signed-off-by: Johan Stridkvist <johan.stridkvist@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -47,6 +47,13 @@ config BT_RX_BUF_LEN
 	  an L2CAP MTU of 65 bytes. On top of this there's the L2CAP header
 	  (4 bytes) and the ACL header (also 4 bytes) which yields 73 bytes.
 
+config BT_BUF_ACL_DATA_LEN
+    int
+	range 27 251
+	default 65
+	depends on BT_HCI_RAW
+	prompt "Override the maximum ACL data payload." if BT_HCI_RAW
```
Please use the same consistent indentation style as the other option use. Here you have everything after `int` using a different indentation.

jhedberg reviewed (changes requested):
Looks generally fine. One inline comment and then you also have a typo in the commit summary (`_BT` is missing)

joerchan commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -47,6 +47,13 @@ config BT_RX_BUF_LEN
+	range 27 251
+	default 65
+	depends on BT_HCI_RAW
+	prompt "Override the maximum ACL data payload." if BT_HCI_RAW
```
I don't think the if BT_HCI_RAW is needed here when you have depends on BT_HCI_RAW.

joerchan commented on include/bluetooth/hci_raw.h:
```diff
@@ -28,7 +28,12 @@ extern "C" {
+#define BT_BUF_ACL_SIZE MAX(BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU), \
+	(CONFIG_BT_BUF_ACL_DATA_LEN + BT_HCI_ACL_HDR_SIZE + BT_BUF_RESERVE))
+#else
 #define BT_BUF_ACL_SIZE BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU)
```
Just curious, when would this apply? You include hci_raw.h but don't enable BT_HCI_RAW. Could we reduce the amount of possibilities for the BT_BUF_ACL_SIZE and end up with less surprises?

joerchan replied:
Maybe we could eliminate the define above? The ACL shouldn't have to depend on the L2CAP MTU, because that should be fragmented across the LL PDUs. The ACL should be aligned with the LL PDU size like it is when CONFIG_BT_CTLR_TX_BUFFER_SIZE is defined.

It would be good if we could make the buffer size configurations simpler because we end up with many of these situations where depending on the board selected we get different buffer sizes.

jostn replied:
> @jostn Could you have a look at my questions, so that we can proceed?



> @jostn Could you have a look at my questions, so that we can proceed?

I've reworked the PR so that we do not depend on l2cap any longer.

jostn committed:
Bluetooth: HCI: Add option CONFIG_BT_HCI_ACL_DATA_SIZE

Add the option CONFIG_BT_HCI_ACL_DATA_SIZE which allows the user
to configure the max HCI ACL data payload. This is needed for platorms
where the BLE LL, HCI uart and host run on three different cores.

Fixes: #30441

Signed-off-by: Johan Stridkvist <johan.stridkvist@nordicsemi.no>

joerchan commented:
@jostn Could you have a look at my questions, so that we can proceed?

joerchan commented on include/bluetooth/hci_raw.h:
```diff
@@ -21,23 +21,23 @@
+#ifdef CONFIG_BT_HCI_ACL_DATA_SIZE
+#define BT_BUF_ACL_SIZE _BT_ACL_BUF_SIZE(CONFIG_BT_HCI_ACL_DATA_SIZE)
+#else
+#define BT_BUF_ACL_SIZE _BT_ACL_BUF_SIZE(27)
```
CONFIG_BT_HCI_ACL_DATA_SIZE should always be defined, since it only depends on CONFIG_BT_HCI_RAW, and it has the same default value of 27.
Removing this else would mean that the value always comes from Kconfig, which would be cleaner.
